File: multi-agent-pathfinding/js/astar.js
/**
 * astar.js — A* Pathfinding Engine
 * Provides both standard A* and space-time A* (for MAPF conflict avoidance).
 * Space-time A* nodes = (x, y, t) so agents can plan around reservations.
 */

'use strict';

// ---------------------------------------------------------------------------
// MinHeap — generic binary min-heap keyed by a numeric priority
// ---------------------------------------------------------------------------
class MinHeap {
  constructor() {
    this._data = [];
  }

  get size() { return this._data.length; }
  isEmpty() { return this._data.length === 0; }

  push(item) {
    this._data.push(item);
    this._bubbleUp(this._data.length - 1);
  }

  pop() {
    if (this.isEmpty()) return null;
    const top = this._data[0];
    const last = this._data.pop();
    if (this._data.length > 0) {
      this._data[0] = last;
      this._siftDown(0);
    }
    return top;
  }

  peek() { return this._data[0] ?? null; }

  _bubbleUp(i) {
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this._data[parent].f <= this._data[i].f) break;
      [this._data[parent], this._data[i]] = [this._data[i], this._data[parent]];
      i = parent;
    }
  }

  _siftDown(i) {
    const n = this._data.length;
    while (true) {
      let smallest = i;
      const l = 2 * i + 1, r = 2 * i + 2;
      if (l < n && this._data[l].f < this._data[smallest].f) smallest = l;
      if (r < n && this._data[r].f < this._data[smallest].f) smallest = r;
      if (smallest === i) break;
      [this._data[smallest], this._data[i]] = [this._data[i], this._data[smallest]];
      i = smallest;
    }
  }
}

// ---------------------------------------------------------------------------
// AStar — pathfinding on a Grid instance
// ---------------------------------------------------------------------------
class AStar {
  /**
   * @param {Grid} grid
   */
  constructor(grid) {
    this.grid = grid;
  }

  /**
   * Manhattan distance heuristic.
   */
  heuristic(ax, ay, bx, by) {
    return Math.abs(ax - bx) + Math.abs(ay - by);
  }

  /**
   * Standard A* — no time dimension, no reservation table.
   * Returns array of {x, y} from start (exclusive) to goal (inclusive),
   * or null if unreachable.
   */
  findPath(sx, sy, gx, gy) {
    const grid = this.grid;
    if (!grid.isWalkable(gx, gy)) return null;

    const openSet = new MinHeap();
    const cameFrom = new Map();
    const gScore = new Map();

    const startKey = `${sx},${sy}`;
    gScore.set(startKey, 0);
    openSet.push({ x: sx, y: sy, f: this.heuristic(sx, sy, gx, gy), g: 0 });

    while (!openSet.isEmpty()) {
      const curr = openSet.pop();
      if (curr.x === gx && curr.y === gy) return this._reconstructPath(cameFrom, curr.x, curr.y);

      for (const [nx, ny] of grid.getNeighbors(curr.x, curr.y)) {
        const tentativeG = curr.g + 1;
        const nKey = `${nx},${ny}`;
        if (tentativeG < (gScore.get(nKey) ?? Infinity)) {
          gScore.set(nKey, tentativeG);
          cameFrom.set(nKey, { x: curr.x, y: curr.y });
          const h = this.heuristic(nx, ny, gx, gy);
          // Tie-break: prefer nodes closer to goal (smaller h)
          openSet.push({ x: nx, y: ny, f: tentativeG + h + h * 0.001, g: tentativeG });
        }
      }
    }
    return null; // No path
  }

  /**
   * Space-Time A* — plans around a reservation table.
   * Nodes are (x, y, t). Agents can "wait" in place (stay = move to same cell).
   *
   * @param {number} sx - start x
   * @param {number} sy - start y
   * @param {number} gx - goal x
   * @param {number} gy - goal y
   * @param {ReservationTable} reservations
   * @param {number} startTime - timestep when agent departs
   * @param {number} maxTime - maximum timestep horizon
   * @returns {Array<{x,y}>|null} path (t=startTime...) or null
   */
  findPathSpaceTime(sx, sy, gx, gy, reservations, startTime = 0, maxTime = 300) {
    const grid = this.grid;
    if (!grid.isWalkable(gx, gy)) return null;

    // Key: "x,y,t"
    const openSet = new MinHeap();
    const cameFrom = new Map();  // key → {x, y, t}
    const gScore = new Map();

    const startKey = `${sx},${sy},${startTime}`;
    gScore.set(startKey, 0);
    const h0 = this.heuristic(sx, sy, gx, gy);
    openSet.push({ x: sx, y: sy, t: startTime, g: 0, f: h0 });

    while (!openSet.isEmpty()) {
      const curr = openSet.pop();
      const { x, y, t } = curr;

      // Reached goal: reconstruct
      if (x === gx && y === gy) {
        return this._reconstructPathST(cameFrom, x, y, t, startTime);
      }

      if (t >= maxTime) continue;

      // Expand: 4-directional moves + wait-in-place
      const moves = [...grid.getNeighbors(x, y), [x, y]]; // wait = [x,y]
      for (const [nx, ny] of moves) {
        const nt = t + 1;
        // Check vertex reservation at (nx, ny, nt)
        if (reservations.isVertexReserved(nx, ny, nt)) continue;
        // Check edge reservation: swapping positions
        if (reservations.isEdgeReserved(x, y, nx, ny, t)) continue;

        const tentativeG = curr.g + 1;
        const nKey = `${nx},${ny},${nt}`;
        if (tentativeG < (gScore.get(nKey) ?? Infinity)) {
          gScore.set(nKey, tentativeG);
          cameFrom.set(nKey, { x, y, t });
          const h = this.heuristic(nx, ny, gx, gy);
          openSet.push({ x: nx, y: ny, t: nt, g: tentativeG, f: tentativeG + h });
        }
      }
    }
    return null;
  }

  _reconstructPath(cameFrom, x, y) {
    const path = [];
    let key = `${x},${y}`;
    while (cameFrom.has(key)) {
      path.push({ x, y });
      const prev = cameFrom.get(key);
      x = prev.x; y = prev.y;
      key = `${x},${y}`;
    }
    return path.reverse();
  }

  _reconstructPathST(cameFrom, x, y, t, startTime) {
    const path = [];
    let key = `${x},${y},${t}`;
    while (cameFrom.has(key)) {
      path.unshift({ x, y, t });
      const prev = cameFrom.get(key);
      x = prev.x; y = prev.y; t = prev.t;
      key = `${x},${y},${t}`;
    }
    // Include start position at startTime
    path.unshift({ x, y, t: startTime });
    return path;
  }
}

// ---------------------------------------------------------------------------
// ReservationTable — tracks which (x,y,t) cells are claimed by agents
// ---------------------------------------------------------------------------
class ReservationTable {
  constructor() {
    this._vertices = new Map();  // "x,y,t" → agentId
    this._edges    = new Map();  // "x1,y1,x2,y2,t" → agentId
  }

  clear() { this._vertices.clear(); this._edges.clear(); }

  /**
   * Reserve all cells in a space-time path for an agent.
   * Also reserves the goal cell for all future timesteps up to maxTime
   * so agents don't path through it later.
   */
  reservePath(path, agentId, maxTime = 300) {
    if (!path || path.length === 0) return;
    for (let i = 0; i < path.length; i++) {
      const { x, y, t } = path[i];
      this._vertices.set(`${x},${y},${t}`, agentId);
      if (i < path.length - 1) {
        const { x: nx, y: ny, t: nt } = path[i + 1];
        // Forward edge: agent moves from (x,y) to (nx,ny) at time t
        this._edges.set(`${x},${y},${nx},${ny},${t}`, agentId);
        // Reverse edge: block the swap
        this._edges.set(`${nx},${ny},${x},${y},${t}`, agentId);
      }
    }
    // Reserve goal indefinitely to avoid other agents walking through
    const last = path[path.length - 1];
    for (let tt = last.t + 1; tt <= maxTime; tt++) {
      this._vertices.set(`${last.x},${last.y},${tt}`, agentId);
    }
  }

  isVertexReserved(x, y, t) {
    return this._vertices.has(`${x},${y},${t}`);
  }

  isEdgeReserved(fromX, fromY, toX, toY, t) {
    return this._edges.has(`${fromX},${fromY},${toX},${toY},${t}`);
  }

  whoOwns(x, y, t) {
    return this._vertices.get(`${x},${y},${t}`) ?? null;
  }

  /** Remove all reservations for a given agent (for replanning). */
  freeAgent(agentId) {
    for (const [k, v] of this._vertices) if (v === agentId) this._vertices.delete(k);
    for (const [k, v] of this._edges)    if (v === agentId) this._edges.delete(k);
  }
}

File: mapf_ros2_ws/src/mapf_ros/mapf_ros/core/astar.py
"""
astar.py — Space-Time A* + Reservation Table (Python)
Mirrors the JS implementation: MinHeap + standard A* + space-time variant.
"""

from __future__ import annotations
import heapq
from typing import Dict, List, Optional, Set, Tuple


# ── Reservation Table ──────────────────────────────────────────────────────

class ReservationTable:
    """Space-time reservation table shared by all agents during PBS planning."""

    def __init__(self) -> None:
        self._vertices: Dict[Tuple[int,int,int], int] = {}  # (x,y,t) → agent_id
        self._edges:    Dict[Tuple[int,int,int,int,int], int] = {}  # (x1,y1,x2,y2,t) → agent_id

    def clear(self) -> None:
        self._vertices.clear()
        self._edges.clear()

    def reserve_path(self, path: List[dict], agent_id: int, max_time: int = 400) -> None:
        """Reserve all (x,y,t) triples along a space-time path."""
        if not path:
            return
        for i, node in enumerate(path):
            x, y, t = node["x"], node["y"], node["t"]
            self._vertices[(x, y, t)] = agent_id
            if i < len(path) - 1:
                nx, ny, nt = path[i+1]["x"], path[i+1]["y"], path[i+1]["t"]
                # Forward edge
                self._edges[(x, y, nx, ny, t)] = agent_id
                # Block the swap (edge conflict)
                self._edges[(nx, ny, x, y, t)] = agent_id
        # Reserve goal indefinitely to block others from passing through
        last = path[-1]
        for tt in range(last["t"] + 1, max_time + 1):
            self._vertices[(last["x"], last["y"], tt)] = agent_id

    def is_vertex_reserved(self, x: int, y: int, t: int) -> bool:
        return (x, y, t) in self._vertices

    def is_edge_reserved(self, fx: int, fy: int, tx: int, ty: int, t: int) -> bool:
        return (fx, fy, tx, ty, t) in self._edges

    def free_agent(self, agent_id: int) -> None:
        """Remove all reservations belonging to agent_id (for replanning)."""
        self._vertices = {k: v for k, v in self._vertices.items() if v != agent_id}
        self._edges    = {k: v for k, v in self._edges.items()    if v != agent_id}

    def who_owns(self, x: int, y: int, t: int) -> Optional[int]:
        return self._vertices.get((x, y, t))


# ── A* Engine ─────────────────────────────────────────────────────────────

class AStar:
    """A* pathfinder on a Grid instance.

    Provides both standard A* (no time) and space-time A* (with reservations).
    """

    def __init__(self, grid) -> None:
        self.grid = grid

    @staticmethod
    def _heuristic(ax: int, ay: int, bx: int, by: int) -> int:
        return abs(ax - bx) + abs(ay - by)

    def find_path(self, sx: int, sy: int, gx: int, gy: int) -> Optional[List[Tuple[int,int]]]:
        """Standard A* without time dimension.

        Returns list of (x, y) from start (exclusive) to goal (inclusive), or None.
        """
        grid = self.grid
        if not grid.is_walkable(gx, gy):
            return None

        # heap entries: (f, g, x, y)
        open_heap: List[Tuple] = []
        heapq.heappush(open_heap, (self._heuristic(sx, sy, gx, gy), 0, sx, sy))

        came_from: Dict[Tuple[int,int], Optional[Tuple[int,int]]] = {(sx, sy): None}
        g_score:   Dict[Tuple[int,int], int] = {(sx, sy): 0}

        while open_heap:
            _, g, cx, cy = heapq.heappop(open_heap)

            if cx == gx and cy == gy:
                return self._reconstruct(came_from, gx, gy)

            if g > g_score.get((cx, cy), float("inf")):
                continue  # stale entry

            for nx, ny in grid.get_neighbors(cx, cy):
                ng = g + 1
                if ng < g_score.get((nx, ny), float("inf")):
                    g_score[(nx, ny)] = ng
                    came_from[(nx, ny)] = (cx, cy)
                    h = self._heuristic(nx, ny, gx, gy)
                    heapq.heappush(open_heap, (ng + h, ng, nx, ny))

        return None

    def find_path_space_time(
        self,
        sx: int, sy: int,
        gx: int, gy: int,
        reservations: ReservationTable,
        start_time:   int = 0,
        max_time:     int = 400,
    ) -> Optional[List[dict]]:
        """Space-Time A* with reservation avoidance.

        Returns list of {"x", "y", "t"} dicts from start_time to arrival, or None.
        """
        grid = self.grid
        if not grid.is_walkable(gx, gy):
            return None

        # heap: (f, g, x, y, t)
        h0 = self._heuristic(sx, sy, gx, gy)
        open_heap: List[Tuple] = [(h0, 0, sx, sy, start_time)]

        came_from: Dict[Tuple, Optional[Tuple]] = {(sx, sy, start_time): None}
        g_score:   Dict[Tuple, int] = {(sx, sy, start_time): 0}

        while open_heap:
            _, g, cx, cy, ct = heapq.heappop(open_heap)

            if cx == gx and cy == gy:
                return self._reconstruct_st(came_from, cx, cy, ct, start_time)

            if ct >= max_time:
                continue
            if g > g_score.get((cx, cy, ct), float("inf")):
                continue  # stale

            # Expand: neighbours + wait-in-place
            moves = list(grid.get_neighbors(cx, cy)) + [(cx, cy)]
            for nx, ny in moves:
                nt = ct + 1
                # Vertex conflict check
                if reservations.is_vertex_reserved(nx, ny, nt):
                    continue
                # Edge (swap) conflict check
                if reservations.is_edge_reserved(cx, cy, nx, ny, ct):
                    continue

                ng = g + 1
                state = (nx, ny, nt)
                if ng < g_score.get(state, float("inf")):
                    g_score[state] = ng
                    came_from[state] = (cx, cy, ct)
                    h = self._heuristic(nx, ny, gx, gy)
                    heapq.heappush(open_heap, (ng + h, ng, nx, ny, nt))

        return None

    # ── Path reconstruction ────────────────────────────────────────────────

    @staticmethod
    def _reconstruct(came_from: dict, x: int, y: int) -> List[Tuple[int,int]]:
        path = []
        curr: Optional[Tuple] = (x, y)
        while curr is not None:
            path.append(curr)
            curr = came_from.get(curr)
        path.reverse()
        return path[1:]  # exclude start cell

    @staticmethod
    def _reconstruct_st(
        came_from: dict,
        x: int, y: int, t: int,
        start_time: int,
    ) -> List[dict]:
        path = []
        state: Optional[Tuple] = (x, y, t)
        while state is not None:
            path.append({"x": state[0], "y": state[1], "t": state[2]})
            state = came_from.get(state)
        path.reverse()
        return path
